File: algo/problems/0049-group-anagrams/js/groupAnagrams.js
/**
 * 以"排序后的字符串"为键，用 Map 聚合。O(n·k·log k)。
 * @param {string[]} strs
 * @returns {string[][]}
 */
function groupAnagrams(strs) {
  const groups = new Map();
  for (const s of strs) {
    const key = s.split('').sort().join('');
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(s);
  }
  return [...groups.values()];
}

module.exports = { groupAnagrams };

File: algo/problems/0049-group-anagrams/python/group_anagrams.py
from collections import defaultdict


def group_anagrams(strs: list[str]) -> list[list[str]]:
    """以"排序后的字符串"为键聚合。O(n·k·log k)。
    更优键：用 26 维字符计数元组作键，可降到 O(n·k)。"""
    groups: dict[str, list[str]] = defaultdict(list)
    for s in strs:
        key = "".join(sorted(s))
        groups[key].append(s)
    return list(groups.values())
